### PR TITLE
feat: support match annotation

### LINKED ISSUE #161: Implement annotation to support matching routes by HTTP method, headers, and url paramaters
## How could it be?
Examples:
### 1. Match by method
```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    # support matching multiple or single method 
    higress.io/match-method: "GET POST PATCH"
  name: foo
spec:
  ingressClassName: higress  
  rules:
  - http:
      paths:
      - pathType: Exact
        path: "/"
...
```
### 2. Match by headers

format:
```bash
higress.io/<exact|regex|prefix>-match-header-<header key>: "<part or all of the header value>"
```

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    # exact matching
    higress.io/exact-match-header-abc: "123"
    # regex matching
    higress.io/regex-match-header-content-type: "application/(json|xml)"
    # prefix matching
    higress.io/prefix-match-header-user-id: "10086-1"
  name: foo
spec:
  ingressClassName: higress  
  rules:
  - http:
      paths:
      - pathType: Exact
        path: "/"
...
```

### 2. Match by url parameters

format:
```bash
higress.io/<exact|regex|prefix>-match-query-<param key>: "<part or all of the param value>"
```

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    # exact matching
    higress.io/exact-match-query-abc: "123"
    # regex matching
    higress.io/regex-match-query-name: "B.*"
    # prefix matching
    higress.io/prefix-match-query-age: "2"
  name: foo
spec:
  ingressClassName: higress  
  rules:
  - http:
      paths:
      - pathType: Exact
        path: "/"
...
```

## Other related information

[1. Annotation example](https://github.com/alibaba/higress/blob/main/pkg/ingress/kube/annotations/cors.go)
[2. Istio api conversion](https://github.com/alibaba/higress/blob/main/pkg/ingress/config/ingress_config.go#L331)
[3. Istio VitrualService's MatchRequest api](https://istio.io/latest/docs/reference/config/networking/virtual-service/#HTTPMatchRequest)

### PR DESCRIPTION
**Pull request by Charlie17Li**
### Ⅰ. Describe what this PR did

support matching routes by HTTP method, headers, and URL parameters

### Ⅱ. Does this pull request fix one issue?

fixes #161

### CODE REVIEW

File: test/ingress/conformance/tests/httproute-match-headers.yaml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    - host: "foo.com"
+      http:
+        paths:
+          - pathType: Exact
+            path: "/foo"
+            backend:
+              service:
+                name: infra-backend-v1
+                port:
+                  number: 8080
```
Comment (Xunzhuo) [inline test/ingress/conformance/tests/httproute-match-headers.yaml:39]: The last line should need a newline

File: test/ingress/conformance/tests/httproute-match-methods.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    - host: "foo.com"
+      http:
+        paths:
+          - pathType: Exact
+            path: "/foo"
+            backend:
+              service:
+                name: infra-backend-v1
+                port:
+                  number: 8080
```
Comment (Xunzhuo) [inline test/ingress/conformance/tests/httproute-match-methods.yaml:34]: The last line should need a newline

File: test/ingress/conformance/tests/httproute-match-query-params.yaml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    - host: "foo.com"
+      http:
+        paths:
+          - pathType: Exact
+            path: "/foo"
+            backend:
+              service:
+                name: infra-backend-v1
+                port:
+                  number: 8080
```
Comment (Xunzhuo) [inline test/ingress/conformance/tests/httproute-match-query-params.yaml:39]: The last line should need a newline

**Review by Xunzhuo — COMMENTED** [review]
Mostly LGTM, some nits and a question:

Do we need support `nginx.ingress.kubernetes.io` prefix ? cc @johnlanni

File: pkg/ingress/kube/annotations/match.go [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+				route.Match[i].QueryParams = map[string]*networking.StringMatch{}
+			}
+			addQueryParamsMatch(route.Match[i].QueryParams, config)
+			IngressLog.Debug(fmt.Sprintf("match : %v, queryParams: %v", route.Match[i].Name, route.Match[i].QueryParams))
+		}
+	}
+}
+
+func (m match) matchByMethod(annotations Annotations, ingress *Ingress) error {
+	if !annotations.HasASAP(matchMethod) {
```
Comment (johnlanni) [inline pkg/ingress/kube/annotations/match.go:99]:
```suggestion
	if !annotations.HasHigress(matchMethod) {
```
Only the `higress.io` prefix is required.

File: pkg/ingress/kube/annotations/match.go [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+		return nil
+	}
+
+	config := ingress.Match
+	str, err := annotations.ParseStringASAP(matchMethod)
```
Comment (johnlanni) [inline pkg/ingress/kube/annotations/match.go:104]:
```suggestion
	str, err := annotations.ParseStringForHigress(matchMethod)
```

**Review by johnlanni — CHANGES_REQUESTED** [review]

**Comment by johnlanni** [comment]
> Mostly LGTM, some nits and a question:
> 
> Do we need support `nginx.ingress.kubernetes.io` prefix ? cc @johnlanni

Only those annotations compatible with nginx-ingress are required.

**Review by Xunzhuo — APPROVED** [review]
Thanks @Charlie17Li, LGTM after CI passed.

File: pkg/ingress/kube/annotations/match.go [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+			set[t] = struct{}{}
+			config.Methods = append(config.Methods, t)
+		}
+	}
+
+	return nil
+}
+
+// matchByHeader to parse annotations to find matchHeader config
+func (m match) matchByHeader(annotations Annotations, config *Ingress) error {
```
Comment (johnlanni) [inline pkg/ingress/kube/annotations/match.go:123]: Could you reuse the same function for `matchByHeader` and `matchByUrlParam`?

File: pkg/ingress/kube/annotations/match.go [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+			config.Headers[prefix] = make(map[string]string)
+		}
+		config.Headers[prefix][k[start:]] = v
+		return nil
+	}
+
+	return ErrInvalidAnnotationName
+}
+
+func (m match) doMatchQuery(k, v string, ingress *Ingress, start int) error {
```
Comment (johnlanni) [inline pkg/ingress/kube/annotations/match.go:189]: Could you reuse the same function for doMatchHeader and doMatchQuery?

**Review by johnlanni — CHANGES_REQUESTED** [review]

File: pkg/ingress/kube/annotations/match.go [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+	var sb strings.Builder
+	for i := 0; i < len(methods); i++ {
+		if i != 0 {
+			sb.WriteString("|")
+		}
+		sb.WriteString(methods[i])
+	}
+
+	return &networking.StringMatch{
+		MatchType: &networking.StringMatch_Regex{
```
[the remark this answers is not in the document]
Comment (johnlanni) [reply pkg/ingress/kube/annotations/match.go:197]: You are right, I misunderstood.

**Review by johnlanni — APPROVED** [review]
LGTM

**Pull request merged**